### ISSUE TITLE
How to calculate the recall of question-to-column alignment

### ISSUE THREAD

**Issue by zhangjiahui-buaa** [tzshi/squall#2]
In Table 8, recall of q2c, c2q, d2q is given and I am stuck at reproducing the result of it 

Surprisingly, the reproduced result is much better than Table 8. I guess there is something wrong with my implementation.

It would be great if you can provide me with your implementation of the calculation of recall.

Thank you very much

**Comment by tzshi** [comment]
Hi,

For the recalls, we opted for the "soft" version. Say, for example, the binary masks for the gold q2c alignments are `gold_mask`, and the model-generated attention matrix is `attn_q2c`. Recall for this instance would be `np.sum(gold_mask * attn_q2c) / len(question)`. To get recall over the corpus, we summed over all the sums, and then divided that by the total number of tokens.

Hope this clarifies.

I'd also be curious about what metrics, if different, you used, and what the range of recall values you got. Did you still observe the same qualitative results as in our paper, if quantitatively different?

**Comment by zhangjiahui-buaa** [comment]
Thank you for your reply.

Here is core part my metrics: 
~~~python
actual_positive = 0
true_positive  = 0
for each attn_q2c:
    pred_label = argmax(attn_q2c)
        for pred, gold in zip(pred_label, gold_label):
            if gold != -1: # only consider token that has actual alignment
                actual_positive += 1
                if pred == gold:
                    true_positive += 1
~~~
iterate all the questions. Then true_positive/actual_positive is the recall. This reaches about 83% on the development set if encoder is equipped with BERT. Without BERT, it's 73%.

By the way. About your method.

>Recall for this instance would be np.sum(gold_mask * attn_q2c) / len(question)

Here, why the denominator is len(question), instead of np.sum(gold_mask)

**Comment by tzshi** [comment]
Thanks for showing your code for recall calculation. Yours is the "hard" version and we used the "soft" version because mappings can potentially be one-to-many (might not be too much an issue for q2c, but this is mostly for consistency when considering d2q etc.).

For the same reason `np.sum(gold_mask)` can double-count tokens. But you were right that in my previous answer, I omitted the condition that we are only looking at tokens that are aligned.
